packages/ckeditor5-image/src/image/imageediting.js

```diff
@@ -181,8 +176,15 @@ function matchImageInsideParent( parentName ) {
 			return null;
 		}
 
-		// Convert only images inside paragraph.
-		if ( !parent || !parent.is( 'element', parentName ) ) {
+		const isBlockImage = ( parent &&
+			( parent.is( 'element', 'figure' ) ||
+				( parent.is( 'element', 'a' ) && parent.parent.is( 'element', 'figure' ) )
+			)
+		);
+
+		const incorrectParent = matchInlineImage ? isBlockImage : !isBlockImage;
```

Comment:
This is hard to read, please try to break it into a more readable form.

Reply:
Also, I think that it should not check for the `'a'` element, it should find the closest ancestor that is a figure with the required class name ([`findAncestor`](https://ckeditor.com/docs/ckeditor5/latest/api/module_engine_view_element-Element.html#function-findAncestor)).

Reply:
I second that this is hard to read. This should be more like this:

```
if ( matchInlineImage ) {
	if ( ... ) {
		// ...
	}
	
	// ...
	
	if ( ... ) {
		return null;
	}
} else {
	if ( ... ) {
		// ...
	}
	
	// ...
	
	if ( ... ) {
		return null;
	}
}
```

Where did this `'a'` parent check come from BTW?

Reply:
`a` came from link tests :) When image has a link so there is `<figure><a><image></image></a></figure>` then `figure` is no longer a direct ancestor of an `image`